src/Common/Threading.cpp

```diff
@@ -0,0 +1,484 @@
+    }
+
+    inline UInt32 * upperAddress(void * address)
+    {
+        return reinterpret_cast<UInt32 *>(address) + (std::endian::native == std::endian::little);
+    }
+
+    inline void waitLowerFetch(std::atomic<UInt64> & address, UInt64 & value)
+    {
+        futexWait(lowerAddress(&address), lowerValue(value));
+        value = address.load();
+    }
+
+    inline bool cancellableWaitLowerFetch(std::atomic<UInt64> & address, UInt64 & value)
+    {
+        bool res = CancelToken::local().wait(lowerAddress(&address), lowerValue(value));
+        value = address.load();
+        return res;
+    }
+
+    inline void wakeLowerOne(std::atomic<UInt64> & address)
+    {
+        syscall(SYS_futex, lowerAddress(&address), FUTEX_WAKE_PRIVATE, 1, nullptr, nullptr, 0);
+    }
+
+    // inline void wakeLowerAll(std::atomic<UInt64> & address)
+    // {
+    //     syscall(SYS_futex, lowerAddress(&address), FUTEX_WAKE_PRIVATE, INT_MAX, nullptr, nullptr, 0);
+    // }
+
+    inline void waitUpperFetch(std::atomic<UInt64> & address, UInt64 & value)
+    {
+        futexWait(upperAddress(&address), upperValue(value));
+        value = address.load();
+    }
+
+    inline bool cancellableWaitUpperFetch(std::atomic<UInt64> & address, UInt64 & value)
+    {
+        bool res = CancelToken::local().wait(upperAddress(&address), upperValue(value));
+        value = address.load();
+        return res;
+    }
+
+    // inline void wakeUpperOne(std::atomic<UInt64> & address)
+    // {
+    //     syscall(SYS_futex, upperAddress(&address), FUTEX_WAKE_PRIVATE, 1, nullptr, nullptr, 0);
+    // }
+
+    inline void wakeUpperAll(std::atomic<UInt64> & address)
+    {
+        syscall(SYS_futex, upperAddress(&address), FUTEX_WAKE_PRIVATE, INT_MAX, nullptr, nullptr, 0);
+    }
+}
+
+void CancelToken::Registry::insert(CancelToken * token)
+{
+    std::lock_guard<std::mutex> lock(mutex);
+    threads[token->thread_id] = token;
+}
+
+void CancelToken::Registry::remove(CancelToken * token)
+{
+    std::lock_guard<std::mutex> lock(mutex);
+    threads.erase(token->thread_id);
+}
+
+void CancelToken::Registry::signal(UInt64 tid)
+{
+    std::lock_guard<std::mutex> lock(mutex);
+    if (auto it = threads.find(tid); it != threads.end())
+        it->second->signalImpl();
+}
+
+void CancelToken::Registry::signal(UInt64 tid, int code, const String & message)
+{
+    std::lock_guard<std::mutex> lock(mutex);
+    if (auto it = threads.find(tid); it != threads.end())
+        it->second->signalImpl(code, message);
+}
+
+const std::shared_ptr<CancelToken::Registry> & CancelToken::Registry::instance()
+{
+    static std::shared_ptr<Registry> registry{new Registry()}; // shared_ptr is used to enforce correct destruction order of tokens and registry
+    return registry;
+}
+
+CancelToken::CancelToken()
+    : state(disabled)
+    , thread_id(getThreadId())
+    , registry(Registry::instance())
+{
+    registry->insert(this);
```

Comment:
`Registry::instance()->insert(this)`? (same in l. 162) - just for consistency with l. 167/172

Reply:
I've made it for consistency with l.162. because ctor/dtor. but `signal()` is static, so `registry` cannot be used